src/globus_sdk/exc/api.py

```diff
@@ -11,33 +12,67 @@
 
 log = logging.getLogger(__name__)
 
+_CACHE_SENTINEL = object()
+
+
+class _ErrorFormat(enum.Enum):
+    undefined = enum.auto()
+    jsonapi = enum.auto()
+    type_zero = enum.auto()
+
 
 class GlobusAPIError(GlobusError):
     """
     Wraps errors returned by a REST API.
 
-    :ivar http_status: HTTP status code (int)
-    :ivar code: Error code from the API (str),
-                or "Error" for unclassified errors
-    :ivar message: Error message from the API. In general, this will be more
-                   useful to developers, but there may be cases where it's
-                   suitable for display to end users.
+    :ivar int http_status: HTTP status code
+    :ivar str code: Error code from the API or "Error" for unclassified errors
+    :ivar str request_id: The 'request_id' included in the error data, if any.
+    :ivar list[str] messages: A list of error messages, extracted from the response
+        data. If the data cannot be parsed or does not contain any clear message fields,
+        this list may be empty.
+    :ivar list[GlobusSubError] errors: A list of sub-error documents, as would be
+        presented by JSON:API APIs and similar interfaces.
     """
 
-    MESSAGE_FIELDS = ["message", "detail"]
+    MESSAGE_FIELDS = ["message", "detail", "title"]
     RECOGNIZED_AUTHZ_SCHEMES = ["bearer", "basic", "globus-goauthtoken"]
 
     def __init__(self, r: requests.Response, *args: t.Any, **kwargs: t.Any):
+        self._cached_raw_json: t.Any = _CACHE_SENTINEL
+
         self.http_status = r.status_code
         # defaults, may be rewritten during parsing
-        self.code = "Error"
-        self.message = r.text
+        self.code: str | None = "Error"
+        self.request_id: str | None = None
+        self.messages: list[str] = []
+        self.errors: list[ErrorSubdocument] = []
 
         self._info: ErrorInfoContainer | None = None
         self._underlying_response = r
         self._parse_response()
         super().__init__(*self._get_args())
 
+    @property
+    def message(self) -> str | None:
+        """
+        An error message from the API.
+
+        If there are multiple messages available, this will contain all messages joined
+        with semicolons. If there is no message available, this will be ``None``.
+        """
+        if self.messages:
+            return "; ".join(self.messages)
+        return None
+
+    @message.setter
+    def message(self, value: str) -> None:
+        warn_deprecated(
+            "Setting a message on GlobusAPIError objects is deprecated. "
+            "This overwrites any parsed messages. Append to 'messages' instead."
+        )
```

Comment:
No, users should never see this from our first-party globus-sdk behaviors. We are no longer setting `message` directly, so the SDK never triggers this codepath (except in the test for this).

The way it would likely happen is with third-party tooling built on top of the SDK which sets `message`. Or, if the user is maintaining that tooling, it's their own second-party code using globus-sdk.

Consider this one:

```python
class MyError(globus_sdk.GlobusAPIError):
    def __init__(self, r):
        super().__init__(r)
        self.message = "For the lulz"
```

That error class will emit the warning. If the maintainer is behaving as we hope and has some testing environment where they see those warnings (e.g. pytest), they'll see the warning when they test with the new SDK version.